File: haskell/src/DataStructures/SinglyLinkedList.hs
{-|
Module      : DataStructures.SinglyLinkedList
Description : Singly linked list implementation
Stability   : stable

A purely functional singly linked list.
Time: O(1) for prepend, O(n) for other operations
Space: O(n)
-}
module DataStructures.SinglyLinkedList
  ( SinglyLinkedList
  , empty
  , isEmpty
  , prepend
  , append
  , insert
  , removeAt
  , get
  , indexOf
  , contains
  , size
  , fromList
  , toList
  ) where

-- | A singly linked list (essentially Haskell's built-in list)
newtype SinglyLinkedList a = SinglyLinkedList [a]
  deriving (Show, Eq)

-- | Create an empty list
empty :: SinglyLinkedList a
empty = SinglyLinkedList []

-- | Check if list is empty
isEmpty :: SinglyLinkedList a -> Bool
isEmpty (SinglyLinkedList xs) = null xs

-- | Add element to the front
prepend :: a -> SinglyLinkedList a -> SinglyLinkedList a
prepend x (SinglyLinkedList xs) = SinglyLinkedList (x : xs)

-- | Add element to the back
append :: a -> SinglyLinkedList a -> SinglyLinkedList a
append x (SinglyLinkedList xs) = SinglyLinkedList (xs ++ [x])

-- | Insert element at index
insert :: Int -> a -> SinglyLinkedList a -> Maybe (SinglyLinkedList a)
insert i x (SinglyLinkedList xs)
  | i < 0 || i > length xs = Nothing
  | otherwise = Just $ SinglyLinkedList $ take i xs ++ [x] ++ drop i xs

-- | Remove element at index
removeAt :: Int -> SinglyLinkedList a -> Maybe (a, SinglyLinkedList a)
removeAt i (SinglyLinkedList xs)
  | i < 0 || i >= length xs = Nothing
  | otherwise = Just (xs !! i, SinglyLinkedList $ take i xs ++ drop (i + 1) xs)

-- | Get element at index
get :: Int -> SinglyLinkedList a -> Maybe a
get i (SinglyLinkedList xs)
  | i < 0 || i >= length xs = Nothing
  | otherwise               = Just (xs !! i)

-- | Find index of first occurrence
indexOf :: Eq a => a -> SinglyLinkedList a -> Maybe Int
indexOf x (SinglyLinkedList xs) = findIndex' 0 xs
  where
    findIndex' _ []     = Nothing
    findIndex' n (y:ys)
      | x == y    = Just n
      | otherwise = findIndex' (n + 1) ys

-- | Check if element exists
contains :: Eq a => a -> SinglyLinkedList a -> Bool
contains x (SinglyLinkedList xs) = x `elem` xs

-- | Get the number of elements
size :: SinglyLinkedList a -> Int
size (SinglyLinkedList xs) = length xs

-- | Create from list
fromList :: [a] -> SinglyLinkedList a
fromList = SinglyLinkedList

-- | Convert to list
toList :: SinglyLinkedList a -> [a]
toList (SinglyLinkedList xs) = xs

File: go/datastructures/singly_linked_list.go
package datastructures

// SinglyLinkedList implements a singly linked list.
// Time: O(1) front ops, O(n) back ops and search
// Space: O(n)
type SinglyLinkedList[T comparable] struct {
	head *singlyNode[T]
	tail *singlyNode[T]
	size int
}

type singlyNode[T any] struct {
	value T
	next  *singlyNode[T]
}

func NewSinglyLinkedList[T comparable]() *SinglyLinkedList[T] {
	return &SinglyLinkedList[T]{}
}

func (l *SinglyLinkedList[T]) PushFront(value T) {
	node := &singlyNode[T]{value: value, next: l.head}
	l.head = node
	if l.tail == nil {
		l.tail = node
	}
	l.size++
}

func (l *SinglyLinkedList[T]) PushBack(value T) {
	node := &singlyNode[T]{value: value}
	if l.tail != nil {
		l.tail.next = node
	} else {
		l.head = node
	}
	l.tail = node
	l.size++
}

func (l *SinglyLinkedList[T]) PopFront() (T, bool) {
	var zero T
	if l.head == nil {
		return zero, false
	}
	value := l.head.value
	l.head = l.head.next
	if l.head == nil {
		l.tail = nil
	}
	l.size--
	return value, true
}

func (l *SinglyLinkedList[T]) PeekFront() (T, bool) {
	var zero T
	if l.head == nil {
		return zero, false
	}
	return l.head.value, true
}

func (l *SinglyLinkedList[T]) PeekBack() (T, bool) {
	var zero T
	if l.tail == nil {
		return zero, false
	}
	return l.tail.value, true
}

func (l *SinglyLinkedList[T]) Contains(value T) bool {
	current := l.head
	for current != nil {
		if current.value == value {
			return true
		}
		current = current.next
	}
	return false
}

func (l *SinglyLinkedList[T]) Remove(value T) bool {
	if l.head == nil {
		return false
	}
	if l.head.value == value {
		l.head = l.head.next
		if l.head == nil {
			l.tail = nil
		}
		l.size--
		return true
	}
	current := l.head
	for current.next != nil {
		if current.next.value == value {
			if current.next == l.tail {
				l.tail = current
			}
			current.next = current.next.next
			l.size--
			return true
		}
		current = current.next
	}
	return false
}

func (l *SinglyLinkedList[T]) IsEmpty() bool {
	return l.size == 0
}

func (l *SinglyLinkedList[T]) Size() int {
	return l.size
}

func (l *SinglyLinkedList[T]) ToSlice() []T {
	result := make([]T, 0, l.size)
	current := l.head
	for current != nil {
		result = append(result, current.value)
		current = current.next
	}
	return result
}

func (l *SinglyLinkedList[T]) Clear() {
	l.head = nil
	l.tail = nil
	l.size = 0
}
